partiql-ast/README.adoc

```diff
@@ -0,0 +1,364 @@
+        from = table("T")
+    }
+}
+----
+
+.Custom Factory DSL Example
+[source,kotlin]
+----
+import org.partiql.ast.builder.ast
+
+// This will instantiate your custom `ComparableIdentifier`. Nice!
+ast(myFactory) {
+    exprSFW {
+        select = select("x", "y", "z")
+        from = table("T")
+    }
+}
+----
+
+IMPORTANT: The last examples works because the DSL block closes over the factory with an AstBuilder. This means that
+the helper extensions or any DSL usage will use the provided factory!
+
+=== Builder Usage
+
+The DSL is not much more than Kotlin syntactic sugar over traditional fluent-builder classes. If you are coming from Java, these will be useful.
+Every node defines a static `builder()` function. Keeping with the previous example, let's see how we can inject our custom
+factory.
+
+[source,kotlin]
+----
+// instance of default IdentifierSymbolImpl
+val a = Identifier.Symbol.builder()
+         .symbol("HELLO")
+         .caseSensitivity(Identifier.CaseSensitivity.INSENSITIVE)
+         .build() // empty, build with default factory
+
+// instance of ComparableIdentifier
+val b = Identifier.Symbol.builder()
+         .symbol("hello")
+         .caseSensitivity(Identifier.CaseSensitivity.INSENSITIVE)
+         .build(myFactory) // nice!
+
+assert(b == a) // TRUE
+assert(a == b) // !! FALSE !! consider always using the same type of factory
+----
+
+== Visitor and Rewriter
+
+The PartiQL AST is a set of interfaces, so how might we extend these for our own purposes? We do not have pattern matching in Kotlin/Java, so we use the visitor pattern.
+
+The visitor pattern is effectively adding methods to each object with some compile safety. You define a behavior and use the node `accept` the behavior. The visitor provides an additional parameter `ctx: C` which is the equivalent of arguments to each method for your behavior.
+
+[source,kotlin]
+----
+public abstract class AstBaseVisitor<R, C> : AstVisitor<R, C> {
+
+    public override fun visit(node: AstNode, ctx: C): R = node.accept(this, ctx)
+
+    public open fun defaultVisit(node: AstNode, ctx: C): R {
+        for (child in node.children) {
+            child.accept(this, ctx)
+        }
+        return defaultReturn(node, ctx)
+    }
+
+    public abstract fun defaultReturn(node: AstNode, ctx: C): R
+}
+----
+
+For example, let's implement a `toString(case: Case)` function on some basic nodes.
+
+[source,kotlin]
+----
+//
+// Usage:
+//  node.accept(ToString, Case.UPPER) // ~ node.toString(Case.UPPER)
+//
+object ToString :  AstBaseVisitor<String, Case>() {
+
+    override fun defaultVisit(node: AstNode, ctx: Case) = defaultReturn(node, ctx)
+
+    override fun defaultReturn(node: AstNode, ctx: Case): String = when (ctx) {
+        Case.UPPER -> node::class.simpleName.uppercase()
+        Case.LOWER -> node::class.simpleName.lowercase()
+        Case.PASCAL -> node::class.simpleName
+        Case.SNAKE -> snakeCaseHelper(node::class.simpleName)
```

Comment:
OTOH (re previous 2 comments) maybe rename `ToString` completely, since it is far from what one can expect from `toString`?  Maybe `AstClassName`?

Reply:
It's not the ast class name, it's the Java simpleName